contents/c++/rvalue.md

```diff
@@ -0,0 +1,145 @@
+
+### Move Semantics
+The main usage of rvalue references is that it allows us to create *move* constructor and *move* assignments, instead of *copy* constructor and *copy* assignments by default. Since rvalues are typically temporary objects, we can just *move* the value instead of *copying* it, thus helping us to save on memory consumption and improving on the performance on our program as well!
+
+One example implementation of a move constructor and move assignment is:
+```cpp
+Foo(Foo&& other) { // move constructor
+    x = other.x;
+    y = other.y;
+    z = other.z;
+}
+
+Foo& operator=(Foo&& other) { // move assignment
+    x = std::move(other.x);
+    y = std::move(other.y);
+    z = std::move(other.z);
+    return *this;
+}
+```
+
+In this move constructor and assignment, the contents of the `other` parameter is *moved* into the object, and the contents in `other` is destroyed afterwards. No additional memory allocation occurs, and the *move* operation is done quickly by a few assignment of address operations, leading to a more performance and memory efficient program.
+
+Along with Move Semantics in C++11, the [STL library](https://www.geeksforgeeks.org/the-c-standard-template-library-stl/) functions also defines the overloaded *move* functions for its container classes, thus we can take advantage of these Move Semantics by simply supplying rvalues, without the need to redefine the classes ourselves.
+
+## Rvalue anti-pattern
+After learning about the new rvalue references and Move Semantics in C++11, many programmers from the older C++ eras may fall into a trap of the rvalue [anti-pattern](https://en.wikipedia.org/wiki/Anti-pattern).
+
+Consider these 2 class constructors:
+```cpp
+Foo(const std::string& x, const std::string& y) { // copy constructor
+    _x = x;
+    _y = y;
+}
+
+Foo(std::string&& x, std::string&& y) { // move constructor
+    _x = x;
+    _y = y;
+}
+```
+
+As we can see, the *move* constructor defines each of its parameter to be rvalues. In this case, if we were to call the constructor using a mixture of both, lvalue for `x` and rvalue for `y`, the copy constructor instead will be called! This is because the copy constructor and take both lvalues and rvalues, both the *move* constructor can only take rvalues. As such, we may think we have made use of Move Semantics to optimize our program, but that may not be the case!
+
+1 solution to this problem would be to overload the constructor for each combination of rvalue parameters possible. However, this is not feasible, especially when the function have many parameters, as we would require `2^n` overload functions where `n` is the number of parameters, leading to too much [boilerplate code](https://www.quora.com/What-is-boilerplate-code), reducing code quality, and increasing memory consumption and compiliation time.
```

Comment:
**One** solution to this problem would be to overload the constructor for **every possible** combination of **the** rvalue parameters. However, this is not feasible, especially when the function **has** many parameters, as we would require 2^n overload functions where n is the number of parameters, leading to too much boilerplate code, reducing code quality, and increasing memory consumption and compiliation time.

This sentence is rather long. Should be broken up into separate sentences.